main/src/ca/uwaterloo/flix/language/phase/TreeShaker.scala

```diff
@@ -31,16 +34,137 @@ import ca.uwaterloo.flix.util.Validation._
+      */
+    def firstReachable(defn: Definition.Constant): Boolean = {
+      val global = defn.sym.namespace.isEmpty
+      val noArguments = defn.formals.isEmpty
+      val notSynthetic = !defn.isSynthetic
+
+      global && noArguments && notSynthetic
+    }
+
+    /**
+      * Searches the given expression `expression` for reachable functions.
+      */
+    def visitExp(expression: Expression): Unit =  expression match {
+      case Expression.Ref(sym, tpe, loc) => newRef(sym)
+      case Expression.Lambda(args, body, tpe, loc) => visitExp(body)
+      case Expression.MkClosure(lambda, freeVars, tpe, loc) => visitExp(lambda)
+      case Expression.MkClosureRef(ref, freeVars, tpe, loc) => visitExp(ref)
+      case Expression.ApplyRef(sym, args, tpe, loc) =>
+        newRef(sym)
+        args.foreach(e => visitExp(e))
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        newRef(sym)
+        actuals.foreach(e => visitExp(e))
+      case Expression.Apply(exp, args, tpe, loc) =>
+        visitExp(exp)
+        args.foreach(e => visitExp(e))
+      case Expression.Unary(op, exp, tpe, loc) => visitExp(exp)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) =>
+        visitExp(exp1)
+        visitExp(exp2)
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+        visitExp(exp1)
+        visitExp(exp2)
+        visitExp(exp3)
+      case Expression.Let(sym, exp1, exp2, tpe, loc) =>
+        visitExp(exp1)
+        visitExp(exp2)
+      case Expression.Is(exp, tag, loc) => visitExp(exp)
+      case Expression.Tag(sym, tag, exp, tpe, loc) => visitExp(exp)
+      case Expression.Untag(tag, exp, tpe, loc) => visitExp(exp)
+      case Expression.Index(base, offset, tpe, loc) => visitExp(base)
+      case Expression.Tuple(elms, tpe, loc) => elms.foreach(e => visitExp(e))
+      case Expression.Existential(fparam, exp, loc) => visitExp(exp)
+      case Expression.Universal(fparam, exp, loc) => visitExp(exp)
+      case _ =>
+    }
+
+    /**
+      * Adds the function `sym` to the map of reachable functions.
+      */
+    def newRef(sym: Symbol.DefnSym): Unit = {
+      reachableFunctions.get(sym) match {
+        // If `sym` has not already been determined reachable, look up its definition in `root`.
+        case None =>
+          root.definitions.get(sym) match {
+            case Some(defn) =>
+              reachableFunctions.put(sym, defn)
+              queue.enqueue(defn)
+            // If `sym` is not defined in `root`, leave this for error checking later.
+            case _ =>
+          }
+        // If `sym` has already been determined reachable, do nothing.
+        case _ =>
+      }
+    }
+
+    /*
+     * We can now use these helper functions to perform tree shaking.
+     */
+
+    // Start the timer.
+    val t = System.nanoTime()
+
+    /*
+     * Find all initially reachable function definitions.
+     */
+    for ((sym, defn) <- root.definitions) {
```

Comment:
So yeah, this is where you *also* need to iterate through:

1. root.constraints
2. root.lattices